api/src/vfs/program.rs

```diff
@@ -153,4 +171,59 @@ impl VirtualNodeHandler for ProgramSourceNodeHandler {
         .select(user_programs::dsl::file_name)
         .get_results(context.conn())?)
     }
+
+    fn set_content(
+        &self,
+        context: &Context,
+        path_variables: &PathVariables,
+        path_segment: &str,
+        content: &str,
+    ) -> Result<()> {
+        use self::user_programs::dsl::*;
+
+        let hw_spec_slug = path_variables.get_var("hw_spec_slug");
+        let program_spec_slug = path_variables.get_var("program_spec_slug");
+
+        // Diesel doesn't support updates with joins, so we have to fetch the
+        // ID of the UserProgram first, then do the update in a second query.
+        let user_program_id: i32 = UserProgram::filter_by_file_name(
+            context.user.id,
+            hw_spec_slug,
+            program_spec_slug,
+            path_segment,
+        )
+        .select(id)
+        .get_result(context.conn())?;
```

Comment:
Can you do subqueries or with clauses so its still executed as one query?

Reply:
idk but I don't think it's worthing trying to figure out. I don't think the performance impact is going to be much at all, the SELECT will be fast.